### PR TITLE
Added BlockState sensitive version of getJumpFactor and getSpeedFactor

### PR DESCRIPTION
**Pull request by ZygZagGaming**
Pr'd so that modders can make blocks influence player/mob speed/jump height based on block states, as well as surroundings if necessary

### CODE REVIEW

File: patches/minecraft/net/minecraft/world/entity/Entity.java.patch [hunk 1]
```diff
@@ -135,6 +135,27 @@
...
           }
        }
+@@ -740,16 +_,16 @@
+    }
+ 
+    protected float m_20098_() {
+-      float f = this.f_19853_.m_8055_(this.m_20183_()).m_60734_().m_49964_();
+-      float f1 = this.f_19853_.m_8055_(this.m_20099_()).m_60734_().m_49964_();
++      float f = this.f_19853_.m_8055_(this.m_20183_()).getJumpFactor(this.f_19853_, this.m_20183_(), this);
++      float f1 = this.f_19853_.m_8055_(this.m_20099_()).getJumpFactor(this.f_19853_, this.m_20183_(), this);
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/entity/Entity.java.patch]: Wrong block position is provided.

File: patches/minecraft/net/minecraft/world/entity/Entity.java.patch [hunk 2]
```diff
@@ -135,6 +135,27 @@
...
+       return (double)f == 1.0D ? f1 : f;
+    }
+ 
+    protected float m_6041_() {
+       BlockState blockstate = this.f_19853_.m_8055_(this.m_20183_());
+-      float f = blockstate.m_60734_().m_49961_();
++      float f = blockstate.getSpeedFactor(this.f_19853_, this.m_20183_(), this);
+       if (!blockstate.m_60713_(Blocks.f_49990_) && !blockstate.m_60713_(Blocks.f_50628_)) {
+-         return (double)f == 1.0D ? this.f_19853_.m_8055_(this.m_20099_()).m_60734_().m_49961_() : f;
++         return (double)f == 1.0D ? this.f_19853_.m_8055_(this.m_20099_()).getSpeedFactor(this.f_19853_, this.m_20183_(), this) : f;
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/entity/Entity.java.patch]: Wrong block position is provided.

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java [hunk 3]
```diff
@@ -920,4 +920,30 @@ default MaterialColor getMapColor(BlockState state, BlockGetter level, BlockPos
     {
         return defaultColor;
     }
+
+    /**
+     * Returns the factor by which jump height is multiplied for any entities standing on this block.
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java]: This documentation doesn't really explain much because there is some additional work going on here. It depends on the jump factor of another block to determine which one is used.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java (hunk 3 above)
Comment (ZygZagGaming) [reply src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:925]: That's not true. The logic you mention is in the Entity class.

Comment (ZygZagGaming) [reply src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java]: Thanks, didn't really think about the secondary logic the Entity class uses

File: patches/minecraft/net/minecraft/world/entity/Entity.java.patch (hunk 1 above)
Comment (ZygZagGaming) [reply patches/minecraft/net/minecraft/world/entity/Entity.java.patch]: Thanks, fixed

File: patches/minecraft/net/minecraft/world/entity/Entity.java.patch [hunk 4]
```diff
@@ -135,6 +135,29 @@
...
        }
+@@ -740,16 +_,18 @@
+    }
+ 
+    protected float m_20098_() {
+-      float f = this.f_19853_.m_8055_(this.m_20183_()).m_60734_().m_49964_();
+-      float f1 = this.f_19853_.m_8055_(this.m_20099_()).m_60734_().m_49964_();
++      float f = this.f_19853_.m_8055_(this.m_20183_()).getJumpFactor(this.f_19853_, this.m_20183_(), this);
++      BlockPos p = this.m_20099_();
++      float f1 = this.f_19853_.m_8055_(p).getJumpFactor(this.f_19853_, p, this);
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/entity/Entity.java.patch]: This can be one line, the computation for this is negligible.

File: patches/minecraft/net/minecraft/world/entity/Entity.java.patch [hunk 5]
```diff
@@ -135,6 +135,29 @@
...
+    }
+ 
+    protected float m_6041_() {
+       BlockState blockstate = this.f_19853_.m_8055_(this.m_20183_());
+-      float f = blockstate.m_60734_().m_49961_();
++      float f = blockstate.getSpeedFactor(this.f_19853_, this.m_20183_(), this);
+       if (!blockstate.m_60713_(Blocks.f_49990_) && !blockstate.m_60713_(Blocks.f_50628_)) {
+-         return (double)f == 1.0D ? this.f_19853_.m_8055_(this.m_20099_()).m_60734_().m_49961_() : f;
++         BlockPos p = this.m_20099_();
++         return (double)f == 1.0D ? this.f_19853_.m_8055_(p).getSpeedFactor(this.f_19853_, p, this) : f;
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/entity/Entity.java.patch]: This can be one line, the computation for this is negligible.

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java [hunk 6]
```diff
@@ -920,4 +920,30 @@ default MaterialColor getMapColor(BlockState state, BlockGetter level, BlockPos
     {
         return defaultColor;
     }
+
+    /**
+     * Returns the factor by which jump height is multiplied for any entities inside this block or standing on this block (unless overridden by the block the entity is in).
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java]: Make this multiline, javadocs are smart enough to know how to interpret it. My general statement is that it should wrap after half a screen.

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java [hunk 7]
```diff
@@ -920,4 +920,30 @@ default MaterialColor getMapColor(BlockState state, BlockGetter level, BlockPos
...
+     * @param pos The block's position in the level
+     * @param entity The entity for which this jump factor is applied
+     */
+    default float getJumpFactor(BlockState state, BlockGetter level, BlockPos pos, Entity entity)
+    {
+        return self().getJumpFactor();
+    }
+
+    /**
+     * Returns the factor by which movement speed is multiplied for any entities inside this block or standing on this block (unless overridden by the block the entity is in).
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java]: Make this multiline, javadocs are smart enough to know how to interpret it. My general statement is that it should wrap after half a screen.

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java [hunk 8]
```diff
@@ -749,4 +749,28 @@ default boolean canBeHydrated(BlockGetter getter, BlockPos pos, FluidState fluid
     {
         return self().getBlock().canBeHydrated(self(), getter, pos, fluid, fluidPos);
     }
+
+    /**
+     * Returns the factor by which movement speed is multiplied for an entity standing on this block.
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java]: Documentation does not match block.

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java [hunk 9]
```diff
@@ -749,4 +749,28 @@ default boolean canBeHydrated(BlockGetter getter, BlockPos pos, FluidState fluid
...
+     * @param pos The block's position in the level
+     * @param entity The entity for which this speed factor is applied
+     */
+    default float getSpeedFactor(BlockGetter getter, BlockPos pos, Entity entity)
+    {
+        return self().getBlock().getSpeedFactor(self(), getter, pos, entity);
+    }
+
+    /**
+     * Returns the factor by which jump height is multiplied for an entity standing on this block.
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java]: Documentation does not match block.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java (hunk 8 above)
Comment (ZygZagGaming) [reply src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java]: What do you mean?

Comment (ChampionAsh5357) [reply src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java]: The documentation is different from that provided on the delegated method.

**Review by ChampionAsh5357 — APPROVED** [review]

**Comment by ChampionAsh5357** [comment]
Forge is moving away from position aware extension methods since the need for block entity based properties is no longer a limiting factor.

As such, the methods should be updated to remove the position and world. The entity can be kept, but there needs to be some justification for it, and probably made nullable for cases where it is not present in mod facing implementations.

**Comment by alcatrazEscapee** [comment]
> some justification for it

An example justification: currently, both speed and jump factor in vanilla affect the entity in a very specific hitbox to when they interact with entities. (This behavior causes issues in vanilla, as seen in [MC-182442](https://bugs.mojang.com/browse/MC-182442).)

If a modded block wants to apply a speed or jump factor, and *not* run into this issue, they need to essentially know both (a) the position of the entity, and (b), the position of the block in question being touched, in order to be able to return a jump or slowing factor appropriately. Without both of these, checking if the entity is *actually* near the block's specific region of slowing/jump-factoring can be difficult.

In short, these methods are used by vanilla to apply entity-block collision effects. Knowing the entity and block in question are useful for precise application of these factors. You cannot collapse this into block state properties, nor is this a "block entity based property".

**Comment by ZygZagGaming** [comment]
> Forge is moving away from position aware extension methods since the need for block entity based properties is no longer a limiting factor.

What changed?

**Comment by Technici4n** [comment]
> Forge is moving away from position aware extension methods since the need for block entity based properties is no longer a limiting factor.

Not sure I understand. What about blocks such as AE2 multipart cables that can't encode all the information in a block state? These hooks need to stay...

**Comment by ChampionAsh5357** [comment]
So, in general, the typical usage of position aware block hooks are to mimic the behavior of other blocks. However, a good portion of maintenance for these hooks are difficult and prevent these properties from being properly used outside of any position aware context, such as world generation as certain data may not be available at certain times. As such, the current stance is that, unless there is a valid justification, new position aware hooks will not be allowed, and old position aware hooks may be removed once their patch location significantly changes in a future version if it doesn't meet the justification burden.

If there is a valid reason why jump and speed factor need to be dynamically adjusted from a block entity that couldn't be attempted from the state or a separate block, then please provide one. So far, I have seen no reason why a block entity should effect how an entity moves on the block. Even blocks with multidirectional applications can be achieved with a direction block state.

**Comment by eerussianguy** [comment]
> So far, I have seen no reason why a block entity should effect how an entity moves on the block

And yet this is completely ignoring that level and pos solves problems completely unrelated to block entities. Just ignoring that does not strengthen your argument here.

**Comment by ChampionAsh5357** [comment]
> And yet this is completely ignoring that level and pos solves problems completely unrelated to block entities. Just ignoring that does not strengthen your argument here.

What problems? Stating a claim without any evidence just makes this a bickering war, which makes this go nowhere, which just makes people annoyed or mad arbitrarily. Currently, the overlords of the Forge team have decided that this is the direction they have chosen, and complaining will get us nowhere. Instead there must be discussions and reasoning behind every comment, rather than just stating an opinion or assumed claim. 

In response to alcatraz's comments (I apologize for not addressing them earlier), it seems the vanilla bug makes more sense to patch at the source rather than have all modders required to fix it themselves, which would remove the need to expose the level or position.

So, please explain in what way does adding the level and position to the jump or speed factor provide a benefit that could not be covered within the block state given an entity context. The only argument you are trying to make here is in response to the current PR. If you want to discuss the general notion Forge is pushing, you can do so on GitHub or through an issue, though I still think it would just turn into a bickering war as both sides are trying to achieve two different goals.

**Comment by eerussianguy** [comment]
I am not in a position to comment a the length I need to this second, but fixing the vanilla bug would break vanilla behavior (aka slabs over ice) so that is not acceptable. And no I have no position on the “direction of forge” that question is not interesting to what is at hand.

**Comment by ChampionAsh5357** [comment]
> fixing the vanilla bug would break vanilla behavior (aka slabs over ice) so that is not acceptable.

Looking over at the associated issues on the topic in further depth (e.g. [MC-1127](https://bugs.mojang.com/browse/MC-1127), [MC-163952](https://bugs.mojang.com/browse/MC-163952), [MC-174574](https://bugs.mojang.com/browse/MC-174574)), it seems as though [MC-182442](https://bugs.mojang.com/browse/MC-182442) which is the bug mentioned seems to be a Working as Intended (WAI) solution within Minecraft, though it remains unresolved, so that is speculation. The codebase defines the logic for both as if current block the entity is in (meaning slabs, carpets, etc.) does not affect the factor of the entity, then the below block will. While an argument could probably be made to say I don't want my non-full block to be affected by what's underneath, I'm not sure why that wouldn't be solved by making the block slightly effect the factor of the movement / jump by a negligible amount (e.g. 1.0000001).

> I am not in a position to comment a the length I need to this second...And no I have no position on the “direction of forge” that question is not interesting to what is at hand.

Sounds good, if there is anything in what you mention that I believe warrants further discussion, I'll bring it up with the other members and continue from there.

**Comment by alcatrazEscapee** [comment]
> So, in general, the typical usage of position aware block hooks are to mimic the behavior of other blocks.

>  As such, the current stance is that, unless there is a valid justification

To clarify (correct me if I'm wrong): the Forge team does not think this constitutes a valid use case, for these patches to be made, considering the maintenance burden? (i.e. a block which mimics another block, by storing it in a block entity and delegating to that for properties, aka aforementioned AE2 multipart cables).

For mods that add these features, is there a suggested remedy from the Forge team? ("don't make these features" / "make it only a set list of blocks it can replace and enumerate them as states" / "just let it be inconsistent" / "coremod the necessary changes in yourself?")

**Comment by ChampionAsh5357** [comment]
> the Forge team does not think this constitutes a valid use case, for these patches to be made, considering the maintenance burden?

Yes, in addition there was also mention that simulation or off world usage of these properties would be limited in usage since their values could only be inferred properly in a position aware context instead of according to the state of the block and any fully identified data.

> For mods that add these features, is there a suggested remedy from the Forge team?

I believe the line of thinking pushes towards using separate blocks due to the block space no longer being limited. So, probably just making them each their own unique block that can copy the properties of the other block.

**Comment by nikita488** [comment]
> I believe the line of thinking pushes towards using separate blocks due to the block space no longer being limited. So, probably just making them each their own unique block that can copy the properties of the other block.

But how to register a "fake" block for each block that exist in the block registry? I guess you mean we need to create and register our block for each block that we want to copy, so if i want to have a copy of grass block, i will create and register my "fake_grass_block" block. But what if i need to have a copy of each block, like for facades? I need to iterate through each block that exist in registry and register "my" block for it, but i'm not sure where and how to register them for this specific case.

**Comment by LexManos** [comment]
Closing this as it's lacking justification. "I want to mimic every block in every configuration" is not a valid justification for these types of patches anymore, as they were originally introduced because of Mojang's limited state range. It causes issues especially when it comes to properties that are used in psudo-worlds such as during generation. Mods that do that already have to deal with 'good enough' compromises and have had to for the entire life of Minecraft. It is my opinion that the overall cost of maintaining such hooks are not worth it for the single shotgun style use case. We will not be removing already existing hooks until they break. However we will not be accepting new ones without additional justification. 

Now to address the Vanilla bug angle. This one is tricky as breaking vanilla behavior is something we try not to do in Forge. This slight logical inconsistency has become a defacto feature of Minecraft. And there is already a way to address the modder 'I want to override whats below me' request. So this isn't good enough justification.

It is sad that this is a super small patch that *probably* wouldn't break anything in the foreseeable future. As it deals with entities that are already in the world and I can't see any worldgen related things that would care about it. But it's just a general design decision that we have made {following Mojang's lead} so we'd need some convincing justifications to deviate from it.
